src/app/Events/EventTemplates.tsx

```diff
@@ -37,76 +37,222 @@
+    setModalOpen(v => {
+      if (v) {
+        setUploadFile(undefined);
+        setUploadFilename('');
+        setUploading(false);
+      }
+      return !v;
+    });
+  };
+
+  const handleFileChange = (value, filename) => {
+    setFileRejected(false);
+    setUploadFile(value);
+    setUploadFilename(filename);
+  };
+
+  const handleUploadSubmit = () => {
+    if (!uploadFile) {
+      window.console.error('Attempted to submit template upload without a file selected');
+      return;
+    }
+    setUploading(true);
+    context.api.addCustomEventTemplate(uploadFile).subscribe(success => {
+      setUploading(false);
+      if (success) {
+        setUploadFile(undefined);
+        setUploadFilename('');
+        refreshTemplates();
+        setModalOpen(false);
+      }
+    });
+  };
+
+  const handleUploadCancel = () => {
+    setUploadFile(undefined);
+    setUploadFilename('');
+    setModalOpen(false);
+  };
+
+  const handleFileRejected = () => {
+    setFileRejected(true);
+  };
+
+  const handleSort = (event, index, direction) => {
+    setSortBy({ index, direction });
+  };
+
   return (<>
     <Toolbar id="event-templates-toolbar">
       <ToolbarContent>
-        <ToolbarItem>
-          <TextInput name="templateFilter" id="templateFilter" type="search" placeholder="Filter..." aria-label="Event template filter" onChange={setFilterText}/>
-        </ToolbarItem>
+        <ToolbarGroup variant="filter-group">
+          <ToolbarItem>
+            <TextInput name="templateFilter" id="templateFilter" type="search" placeholder="Filter..." aria-label="Event template filter" onChange={setFilterText}/>
+          </ToolbarItem>
+        </ToolbarGroup>
+        <ToolbarGroup variant="icon-button-group">
+          <ToolbarItem>
+            <Button variant="plain" aria-label="add" onClick={handleModalToggle}><PlusIcon /></Button>
+          </ToolbarItem>
+        </ToolbarGroup>
       </ToolbarContent>
     </Toolbar>
-    <Table aria-label="Event Templates table" cells={tableColumns} rows={getTemplates()} actions={actions} variant={TableVariant.compact}>
+    <Table aria-label="Event Templates table"
+      variant={TableVariant.compact}
+      cells={tableColumns}
+      rows={displayTemplates}
+      actionResolver={actionResolver}
+      sortBy={sortBy}
+      onSort={handleSort}
+    >
       <TableHeader />
       <TableBody />
     </Table>
+
+    <Modal
+      isOpen={modalOpen}
+      variant={ModalVariant.large}
+      showClose={true}
+      onClose={handleModalToggle}
+      title="Create Custom Event Template"
+      description="Create a customized event template. This is a specialized XML file with the extension .jfc, typically created using JDK Mission Control, which defines a set of events and their options to configure. Not all customized templates are applicable to all targets - a template may specify a custom application event type, which is only available in targets running the associated application."
```

Comment:
>Not all customized templates are applicable to all targets - a template may specify a custom application event type,

The hyphen here should be changed to an em dash (—), or if we want to stick to ASCII, two hyphens (--).